transformer_engine/pytorch/attention.py

```diff
@@ -5090,19 +5089,15 @@ def forward(
                 output.reshape(batch_size * max_seqlen_q // cp_size, -1).transpose_2d()
                 output = output.reshape(batch_size, max_seqlen_q // cp_size, -1)
             else:
-                output = (
-                    output.view(batch_size, max_seqlen_q // cp_size, -1)
-                    .transpose(0, 1)
-                    .contiguous()
-                )
+                output = output.view(batch_size, max_seqlen_q // cp_size, -1).transpose(0, 1)
         elif qkv_format == "bshd":
             # (bs)hd -> bs(hd)
             output = output.reshape(batch_size, max_seqlen_q // cp_size, -1)
         elif qkv_format == "thd":
             # thd -> t(hd)
             output = output.reshape(output.shape[0], -1)
 
-        return output
+        return output.contiguous()
```

Comment:
Wouldn't this force the other paths to call `.contiguous()` as well? Should we remove the last two changes?

Reply:
I removed `.contiguous()` calls on q/k/v for all paths except for CP, so we need a `.contiguous()` here to ensure the continuity of output.

I list the number of `.contiguous()` calls under all situations before and after this change as below:
1. If input q/k/v is contiguous
  a. Before: 4 (q/k/v/out) for `sbhd`, 0 for `bshd` and `thd`.
  b. After: 0 for all paths.
2. If input q/k/v is contiguous on the last dimension but not others
  a. Before: 4 (q/k/v/out) for `sbhd` and 3 (q/k/v) for `bshd` and `thd`.
  b. After: 1 (out) for all paths.
3. If input q/k/v is not contiguous on the last dimension
  a. Before: 4 (q/k/v/out) for `sbhd` and 3 (q/k/v) for `bshd` and `thd`.
  b. After: 4 (q/k/v in FA, out in our code) for `sbhd`, and 3 (q/k/v in FA) for `bshd` and `thd`.

Though not all the cases would happen, moving the `.contiguous()` calls from q/k/v to out does benefit the above case 2).

Reply:
I guess I wasn't sure if every `.contiguous()` call translates to work. Looks like it doesn't: see [here](https://github.com/pytorch/pytorch/blob/e037bb326fdafca243bdb08023bbef93b29a4513/aten/src/ATen/native/TensorProperties.cpp#L109). Then I agree with your assessment.